### PR TITLE
add category model

### PR DESCRIPTION
**Pull request by nikitaabo**

### CODE REVIEW

File: src/main/java/com/example/spring/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    @PostMapping
+    @Operation(summary = "Create a category", description = "Create a new category")
+    public CategoryResponseDto createCategory(@RequestBody @Valid CategoryDto categoryDto) {
+        return categoryService.save(categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER') || hasRole('ROLE_ADMIN')")
+    @GetMapping
+    @Operation(summary = "Get all categories", description = "Get a list of available categories")
+    public List<CategoryResponseDto> getAll() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/CategoryController.java]:
```suggestion
    public List<CategoryResponseDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/com/example/spring/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        return categoryService.findAll();
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER') || hasRole('ROLE_ADMIN')")
+    @GetMapping("/{id}")
+    @Operation(summary = "Get category", description = "Get a category by id")
+    public CategoryResponseDto getCategoryById(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/CategoryController.java]:
```suggestion
    public CategoryResponseDto getCategoryById(@PathVariable @Positive Long id) {
```

File: src/main/java/com/example/spring/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    public CategoryResponseDto updateCategory(@PathVariable @Positive Long id,
+                                      @RequestBody @Valid CategoryDto categoryDto) {
+        return categoryService.update(id, categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    @Operation(summary = "Delete a category", description = "Delete a category by id")
+    public void deleteCategory(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/CategoryController.java]:
```suggestion
    public void deleteCategory(@PathVariable @Positive Long id) {
```

File: src/main/java/com/example/spring/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        categoryService.deleteById(id);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER') || hasRole('ROLE_ADMIN')")
+    @GetMapping("/{id}/books")
+    @Operation(summary = "Get books", description = "Get books by category id")
+    public List<BookDto> getBooksByCategoryId(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/controller/CategoryController.java]:
```suggestion
    public List<BookDto> getBooksByCategoryId(@PathVariable @Positive Long id) {
```

File: src/main/java/com/example/spring/models/Book.java [hunk 5]
```diff
@@ -33,4 +42,14 @@ public class Book {
     private String description;
     private String coverImage;
     private boolean isDeleted = false;
+    @ManyToMany()
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/models/Book.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/com/example/spring/models/Category.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+@Table(name = "categories")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE categories SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/models/Category.java]: you can add unique also

File: src/main/java/com/example/spring/models/Category.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    private String name;
+    private String description;
+    private boolean isDeleted = false;
+
+    public Category() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/models/Category.java]: use Lombok annotation instead

File: src/main/java/com/example/spring/services/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    public List<CategoryResponseDto> findAll() {
+        return categoryRepository.findAll().stream()
+                .map(categoryMapper::toResponse)
+                .toList();
+    }
+
+    @Override
+    public CategoryResponseDto getById(Long id) {
+        return categoryMapper.toResponse(categoryRepository.findById(id).orElseThrow(
+                () -> new NoSuchElementException("Couldn't find category with id: " + id)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/spring/services/CategoryServiceImpl.java]:
```suggestion
                () -> new EntityNotFoundException("Couldn't find category with id: " + id)
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/resources/db/changelog/changes/07-create-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                  name: description
+                  type: varchar(255)
+                  constraints:
+                    nullable: true
+              - column:
+                  name: is_deleted
+                  type: BIT
+                  defaultValue: false
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/07-create-categories-table.yaml:31]: add one empty line at the end of the file (check other files) ;)

**Review by liliia-ponomarenko — APPROVED** [review]
Awesome 😎

**Pull request merged**